### PR TITLE
MH-12266: Propagate Episode Meta Data Changes to OAI-PMH

### PR DESCRIPTION
**Pull request by krinnewitz**
This patch introduces automatic propagation of meta data or ACL changes of episodes to OAI-PMH. Previously, the user had to re-publish an episode in order for the changes to show up in OAI-PMH. Now, as soon as you change meta data and the ACL of an episode, the change is propagated to OAI-PMH without the need to run the re-publish workflows. 

When testing this, please don't forget to replace your activemq.xml with the one that comes with the patch. You can use this URL to check the contents of the OAI-PMH repository:

http://localhost:8080/oaipmh/default?verb=ListRecords&metadataPrefix=matterhorn-inlined

The ACLs of the episodes can be viewed by visiting the XACML resource using your browser. The URI of the XACML file depends on the the episode ID and can be found in the OAI-PMH records.

After changing meta data in the admin UI, it may take a few seconds for the changes to show up in OAI-PMH.

This work is sponsored by SWITCH.

### CODE REVIEW

**Comment by smarquard** [comment]
@krinnewitz, please complete an ICLA contributor licensing agreement for code contributions to Opencast as I don't see you listed on http://licensing.apereo.org/. @staubesv , there should also be a CLA for SWITCH as the entity contracting this work.

**Comment by krinnewitz** [comment]
@smarquard, there already is a CCLA of my Company (plapadoo). Shouldn't this be sufficient?

**Comment by smarquard** [comment]
@krinnewitz, https://www.apereo.org/licensing/agreements says "Individuals still also need to submit an ICLA, even when their organizations has completed a CCLA" so my understanding is you should also submit one.

@lkiesow is a member of the Apereo licensing group so may be the best person to answer more detailed questions (for example whether the organizational CLA here should be plapadoo or SWITCH or both).

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java [hunk 1]
```diff
@@ -321,4 +330,75 @@ private boolean updateEpisodeCatalog(MediaPackage mp, String repository) throws
     return true;
   }
 
+  public void handleEvent(AssetManagerItem.TakeSnapshot snapshotItem) {
+    //An episode or its' ACL has been updated. Construct the MediaPackage and publish it to OAI-PMH.
```
Comment (smarquard) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java]: Trivial, but "its" here does not have an apostrophe in English (same in the comment below), so "its'" should be "its"

**Comment by krinnewitz** [comment]
@smarquard Thanks for pointing out the issue with the ICLAs. I submitted one, but I think it is still pending since I am not yet listed on http://licensing.apereo.org/
@JamesUoM Thank you for your input. I pushed some unit tests which test the handleEvent() method I added previously.

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java (hunk 1 above)
Comment (krinnewitz) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java]: Thank you, I fixed the spelling :)

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/AssetManagerPermissionsUpdatedEventHandler.java [hunk 2]
```diff
@@ -159,7 +159,6 @@ public void handleEvent(final SeriesItem seriesItem) {
         final String orgId = snapshot.getOrganizationId();
         final Organization organization = organizationDirectoryService.getOrganization(orgId);
         if (organization == null) {
-          logger.warn("Skipping update of episode {} since organization {} is unkown", snapshot.getMediaPackage().getIdentifier(), orgId);
```
Comment (staubesv) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/AssetManagerPermissionsUpdatedEventHandler.java]: Thanks ;-)

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/ConductingEpisodeUpdatedEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    public void stopListening() {
+      this.listening = false;
+      future.cancel(true);
+    }
+
+    @Override
+    public void run() {
+      logger.info("Starting to listen for episode update Messages");
```
Comment (staubesv) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/ConductingEpisodeUpdatedEventHandler.java]: **m**essages

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java [hunk 4]
```diff
@@ -321,4 +330,75 @@ private boolean updateEpisodeCatalog(MediaPackage mp, String repository) throws
...
+    // We must be an administrative user to make a query to the OaiPmhPublicationService
+    final User prevUser = securityService.getUser();
+    final Organization prevOrg = securityService.getOrganization();
+
+    try {
+      securityService.setUser(SecurityUtil.createSystemUser(systemAccount, prevOrg));
+      Query query = QueryBuilder.query().mediaPackageId(snapshotItem.getMediapackage()).build();
+      SearchResult result = oaiPmhPersistence.search(query);
+      if (result.getItems().size() != 1) {
+        logger.info("Unexpected number of items: {}", result.getItems().size());
```
Comment (staubesv) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java]: I guess this should be logged as error

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java [hunk 5]
```diff
@@ -321,4 +330,75 @@ private boolean updateEpisodeCatalog(MediaPackage mp, String repository) throws
...
+
+      // item contains the old MediaPackage with all elements. Since unchanged elements are missing in mp, we have to
+      // merge the two media packages.
+      MediaPackage oldMp = item.getMediaPackage();
+
+      // Merge old, unchanged elements
+      MediaPackageSupport.merge(mp, oldMp, MediaPackageSupport.MergeMode.Skip);
+
+      // Does the media package have a title and track?
+      if (!MediaPackageSupport.isPublishable(mp)) {
```
Comment (staubesv) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java]: Hmm.... after all that work you test whether it made sense to do all the work? Can't this be tested before actually doing anything?

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java [hunk 6]
```diff
@@ -321,4 +330,75 @@ private boolean updateEpisodeCatalog(MediaPackage mp, String repository) throws
...
+        logger.info("Media package does not meet criteria for publication");
```
Comment (staubesv) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java]: Please always include media package ID when logging related to media packages. Without that, the message is useless. Also, consider log level debug (or trace) here.

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java [hunk 7]
```diff
@@ -321,4 +330,75 @@ private boolean updateEpisodeCatalog(MediaPackage mp, String repository) throws
...
+        return;
+      }
+
+      // Update the OAI-PMH persistence with the updated mediapackage
+      oaiPmhPersistence.store(mp, item.getRepository());
+    } catch (DistributionException | MediaPackageException | ServiceRegistryException | OaiPmhDatabaseException
+        | NotFoundException e) {
+      logger.warn(e.getMessage());
```
Comment (staubesv) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java]: Hmm... is this warnings or errors? I tend to say that exceptions here are errors...

File: modules/matterhorn-conductor/src/main/resources/OSGI-INF/events/episode-updated.xml [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    </service>
+    <reference name="securityService" interface="org.opencastproject.security.api.SecurityService"
+            cardinality="1..1" policy="static" bind="setSecurityService"/>
+    <reference name="message-broker-receiver" interface="org.opencastproject.message.broker.api.MessageReceiver"
+            cardinality="1..1" policy="static" bind="setMessageReceiver" />
+    <reference name="oaiPmhUpdatedEventHandler"
+               interface="org.opencastproject.event.handler.OaiPmhUpdatedEventHandler"
+               cardinality="1..1" policy="dynamic" bind="setOaiPmhUpdatedEventHandler"/>
+  </scr:component>
+
```
Comment (staubesv) [inline modules/matterhorn-conductor/src/main/resources/OSGI-INF/events/episode-updated.xml]: Superfluous newlines

File: modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java (hunk 5 above)
Comment (krinnewitz) [inline modules/matterhorn-conductor/src/main/java/org/opencastproject/event/handler/OaiPmhUpdatedEventHandler.java]: Good point, but unfortunately, this check only makes sense after merging the media packages. So this cannot be tested earlier.

**Comment by lkiesow** [comment]
Sorry for the late reply. Was on vacation. All required CLAs have been submitted to Apereo and were received by the licensing group. It can take a bit before all lists are updated since the workflow contains some manual checks…

@JamesUoM, since you already started, can you review and merge this? Please merge it into the `3.x` branch on bitbucket.

**Comment by lkiesow** [comment]
The decision was made to not migrate to Github but stay with Bitbucket and Jira.
Can you please re-file this pull request against the Opencast bitbucket repository.

**Pull request closed without merging**